### Importing the necessary libraries.

In [1]:
import numpy as np
import pandas as pd

### Calling the dataset

In [2]:
data = pd.read_csv("Data/creditcard.csv")

data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [3]:
fraud_data = data[data["Class"] == 1]
not_fraud_data = data[data["Class"] == 0]

In [4]:
fraud_data.shape

(492, 31)

In [5]:
not_fraud_data.shape

(284315, 31)

Not to create too much synthetic data i will just be using a small amount of `not_fraud_data`.

As the number of rows in the `fraud_data` is 492, I will be selecting almost 800-900 of the data from the `not_fraud_data` and then SMOTE technique to balance the data.

In [6]:
not_fraud_data_copy = not_fraud_data.sample(n = 900).reset_index(drop = True).copy()

not_fraud_data_copy.head()

Time        V1        V2        V3        V4        V5        V6  \
0  157091.0 -0.484947  0.978998  1.013995  1.259615 -0.335801  0.018406   
1  134056.0  2.148531 -0.693324 -1.124399 -0.118844 -0.629263 -0.897419   
2   48807.0 -0.435792  0.528066  2.206747  0.330896 -0.767276 -0.446252   
3  149701.0  0.097055  1.049243 -0.217881 -0.443058  0.734170 -0.994291   
4  157614.0 -1.068122  1.220688 -1.332905 -1.040678  0.356768  1.149277   

         V7        V8        V9  ...       V21       V22       V23       V24  \
0  0.228071  0.518821  0.095665  ... -0.357288 -0.928097  0.258572 -0.100778   
1 -0.311006 -0.324745 -0.271587  ... -0.223571  0.157629  0.082648  0.043870   
2  0.010626  0.081187  0.511012  ...  0.085661  0.382782  0.054797  0.750286   
3  1.013543 -0.134708 -0.132011  ... -0.299082 -0.701865  0.167489  0.991762   
4 -0.907781 -5.334153 -0.701338  ... -2.663799  0.506519 -0.206425  0.035019   

        V25       V26       V27       V28  Amount  Class  
0 -0.815692 -1.001943  0.133102  0.151276   30.94      0  
1  0.088290  0.728882 -0.037768 -0.066106    9.20      0  
2 -0.919133  0.133291  0.063743  0.191601   11.50      0  
3 -0.482634  0.088135  0.226254  0.089486    8.99      0  
4 -0.201841 -0.174242  0.064178 -0.045665  199.00      0  

[5 rows x 31 columns]

In [7]:
data_new = pd.concat([fraud_data, not_fraud_data_copy], axis = 0).reset_index(drop = True)

In [8]:
data_new.head()

Time        V1        V2        V3        V4        V5        V6  \
0   406.0 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
1   472.0 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
2  4462.0 -2.303350  1.759247 -0.359745  2.330243 -0.821628 -0.075788   
3  6986.0 -4.397974  1.358367 -2.592844  2.679787 -1.128131 -1.706536   
4  7519.0  1.234235  3.019740 -4.304597  4.732795  3.624201 -1.357746   

         V7        V8        V9  ...       V21       V22       V23       V24  \
0 -2.537387  1.391657 -2.770089  ...  0.517232 -0.035049 -0.465211  0.320198   
1  0.325574 -0.067794 -0.270953  ...  0.661696  0.435477  1.375966 -0.293803   
2  0.562320 -0.399147 -0.238253  ... -0.294166 -0.932391  0.172726 -0.087330   
3 -3.496197 -0.248778 -0.247768  ...  0.573574  0.176968 -0.436207 -0.053502   
4  1.713445 -0.496358 -1.282858  ... -0.379068 -0.704181 -0.656805 -1.632653   

        V25       V26       V27       V28  Amount  Class  
0  0.044519  0.177840  0.261145 -0.143276    0.00      1  
1  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
2 -0.156114 -0.542628  0.039566 -0.153029  239.93      1  
3  0.252405 -0.657488 -0.827136  0.849573   59.00      1  
4  1.488901  0.566797 -0.010016  0.146793    1.00      1  

[5 rows x 31 columns]

In [9]:
data_new.drop("Time", axis = 1, inplace = True)

In [10]:
from sklearn.preprocessing import MinMaxScaler

In [11]:
scaler = MinMaxScaler()
scaled_data_new = pd.DataFrame(scaler.fit_transform(data_new), columns = data_new.columns)

In [12]:
data_new

V1        V2        V3        V4        V5        V6        V7  \
0    -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545 -2.537387   
1    -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823  0.325574   
2    -2.303350  1.759247 -0.359745  2.330243 -0.821628 -0.075788  0.562320   
3    -4.397974  1.358367 -2.592844  2.679787 -1.128131 -1.706536 -3.496197   
4     1.234235  3.019740 -4.304597  4.732795  3.624201 -1.357746  1.713445   
...        ...       ...       ...       ...       ...       ...       ...   
1387  2.123994 -1.737910 -1.169054 -1.749448 -1.005771  0.162769 -1.221305   
1388  1.189437 -0.410653  1.011784 -0.790996 -1.272069 -0.622265 -0.686974   
1389 -1.581829 -0.370789  0.998088  2.872422  0.605812  4.000333  1.827070   
1390 -0.982315  0.937896  0.888907 -1.396951 -0.113067 -0.850507  0.294911   
1391 -1.555708  0.164938  2.748192  3.565297 -1.060292  1.684035 -1.049026   

            V8        V9       V10  ...       V21       V22       V23  \
0     1.391657 -2.770089 -2.772272  ...  0.517232 -0.035049 -0.465211   
1    -0.067794 -0.270953 -0.838587  ...  0.661696  0.435477  1.375966   
2    -0.399147 -0.238253 -1.525412  ... -0.294166 -0.932391  0.172726   
3    -0.248778 -0.247768 -4.801637  ...  0.573574  0.176968 -0.436207   
4    -0.496358 -1.282858 -2.447469  ... -0.379068 -0.704181 -0.656805   
...        ...       ...       ...  ...       ...       ...       ...   
1387  0.016136 -1.317689  1.685569  ... -0.101474 -0.037888  0.121911   
1388  0.114039  1.582981 -0.904442  ...  0.050062  0.345192 -0.020167   
1389  0.725349 -1.795785 -0.027682  ...  0.253518 -0.134950  1.272436   
1390  0.527685 -0.433407 -0.727873  ... -0.111995 -0.523971 -0.042609   
1391  0.900376  0.247021  0.190673  ...  0.182194  1.327675  0.470125   

           V24       V25       V26       V27       V28  Amount  Class  
0     0.320198  0.044519  0.177840  0.261145 -0.143276    0.00      1  
1    -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
2    -0.087330 -0.156114 -0.542628  0.039566 -0.153029  239.93      1  
3    -0.053502  0.252405 -0.657488 -0.827136  0.849573   59.00      1  
4    -1.632653  1.488901  0.566797 -0.010016  0.146793    1.00      1  
...        ...       ...       ...       ...       ...     ...    ...  
1387  0.071719 -0.215302 -0.201306 -0.005830 -0.040674  104.65      0  
1388  0.553219  0.435057 -0.715876  0.088576  0.026681    1.00      0  
1389 -1.438527 -0.415672 -0.169703  0.020916  0.181109  515.79      0  
1390 -0.006177 -0.321730  0.664258  0.111853  0.080977    1.52      0  
1391  0.170865 -0.341608  0.457901  0.063879  0.026641   60.33      0  

[1392 rows x 30 columns]

In [13]:
X = data_new.drop("Class", axis = 1)
y = data_new["Class"]

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

In [16]:
from imblearn.over_sampling import SMOTE

In [17]:
smote = SMOTE(sampling_strategy = "auto", random_state = 42)
X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)

In [18]:
X_train_resampled.shape, y_train_resampled.shape

((1438, 29), (1438,))

In [19]:
y_train_resampled.value_counts()

Class
0    719
1    719
Name: count, dtype: int64

In [20]:
from sklearn.linear_model import LogisticRegression

In [21]:
lr = LogisticRegression(max_iter = 5000)

In [22]:
lr.fit(X_train_resampled, y_train_resampled)

LogisticRegression(max_iter=5000)

In [23]:
pred = lr.predict(X_test)

In [24]:
from sklearn.metrics import accuracy_score

In [25]:
accuracy_score(y_test, pred)

0.956989247311828